### PR TITLE
Fixes to `sunpy.sun` calculations

### PR DESCRIPTION
**Pull request by ayshih**
I discovered that most of the calculations in `sunpy.sun.sun` have been broken for quite a while:

* The Julian century conversion had an off-by-one error (the equations are supposed to use offset from J1900.0, which is noon on 1900 January 0, not 1900 January 1)
* The position angle was broken due to two wrong equations
* The true right ascension and true declination appeared to be completely wrong

Note that I haven't done anything to improve the accuracy of these calculations, but the values now once again match with the IDL counterpart (`get_sun.pro` in SSW).

Also note that our tests didn't catch these bugs because they appear to have been written by trusting the incorrect values spit out by the buggy code.  (This is immediately apparent from some values being unphysical.)  Of course, I have committed the same sin by updating the tests the same way.

### CODE REVIEW

**Comment by ayshih** [comment]
It looks like the tests failed in @wafels's solar de-rotation code because it used the solar diameter from `sunpy.sun` (which had a bug with date conversion) and locally re-calculated everything else (without a bug with date conversion).  Thus, the trusted values in the tests were not quite right, and I added a non-zero absolute tolerance to account for numerical precision.

**Comment by ayshih** [comment]
Test failure is unrelated

Fixes #1402

**Comment by dpshelio** [comment]
Finally someone believes me! I've been trying to find time to fix all that in #1070 - because there are few things that are wrong in ssw... but it's been two years since I started that... 
If someone wants to help I am more than happy...

File: sunpy/sun/sun.py [hunk 1]
```diff
@@ -186,17 +186,15 @@ def true_obliquity_of_ecliptic(t='now'):
...
+    x = np.cos(true_longitude(t))
+    true_ra = np.arctan2(y, x)
+    return Longitude(true_ra.to(u.hourangle))
 
 def true_declination(t='now'):
     """Return the true declination."""
-    result = np.cos(true_longitude(t))
-    result = result * u.deg
-    return Latitude(result)
+    result = np.arcsin(np.sin(true_obliquity_of_ecliptic(t))) * np.sin(apparent_longitude(t))
```
Comment (dpshelio) [inline sunpy/sun/sun.py]: why do you do the arcsin of a sin? shouldn't not be the same that the argument? Are you sure this is right? aren't you putting the parenthesis wrong? Look [my implementation](https://github.com/sunpy/sunpy/pull/1070/files#diff-1d9fb5cf2bfd8d46c960d4218bf8a3b0R374) based on "the :blue_book:"

File: sunpy/time/julian.py [hunk 2]
```diff
@@ -12,13 +12,10 @@ def julian_day(t='now'):
...
-# The number of days between Jan 1 1900 and the Julian reference date of
-# 12:00 noon Jan 1, 4713 BC
-JULIAN_DAY_ON_NOON01JAN1900 = 2415021.0
-
 
 def julian_centuries(t='now'):
-    """Returns the number of Julian centuries since 1900 January 0.5."""
-    DAYS_IN_YEAR = 36525.0
+    """Returns the number of Julian centuries since J1900.0 (noon on 1900 January 0)."""
+    DAYS_IN_JULIAN_CENTURY = 36525.0
```
Comment (dpshelio) [inline sunpy/time/julian.py]: Is it TT or TD? - In my PR I am using what's said in the book - which it changes from equation to equation. So I would suggest to be careful with this particular bit.

**Comment by dpshelio** [comment]
I'm not too happy of this kind of "patching" - but since I've not been done much for a while with this I cannnot really oppose to this. I need a ephemerids spring where few of us work the same day at the same time and I'm sure in one day we can have it done properly.

**Comment by wafels** [comment]
What book are you referring to?  Work on sunpy.sun would benefit from
having clear sources that we can all agree on (SSWIDL code does not
count).  For example, Astronomical Algorithms, 2nd Edition.

http://www.amazon.com/dp/0943396611/?tag=googhydr-20&hvadid=38258950767&hvpos=1t1&hvexid=&hvnetw=g&hvrand=2726920829677631662&hvpone=34.95&hvptwo=&hvqmt=b&hvdev=c&ref=pd_sl_uqp3k0ojg_b

**Comment by dpshelio** [comment]
That's :blue_book: (though it's red) :) - you can see [it mentioned in my PR](https://github.com/sunpy/sunpy/pull/1070/files#diff-8d3099c1c55f1429fb3c97e54fad4063R26)

**Comment by ayshih** [comment]
Fixed the bug that @dpshelio spotted and some other cleanup

Yes, this PR is a "patch" – we still need to do a better job with the ephemeris calculations – but my objective here is to fix implementation bugs (e.g., a misplaced parenthesis that was identified over a year ago!), even if the results are slightly inaccurate.

**Comment by Cadair** [comment]
@ayshih please merge in master or rebase to fix travis.

**Comment by Cadair** [comment]
@ayshih I am going to assume this should be backported? Should it also be backported to 0.6?

**Comment by ayshih** [comment]
Ugh, not a trivial backport, so it'll take a bit more attention from me to put up the backport PR.

**Comment by Cadair** [comment]
I assume you mean to 0.6? or are there major changes to the 0.7 sun.py? If you can do the 0.7 version asap I will throw it into 0.7.1 which I was hoping to do today, but might have to wait till the weekend.

**Comment by ayshih** [comment]
Yes, I was referring to the 0.6 backport.  The cherry-pick had a conflict with my changes to `sunpy.time.julian`, so I just need to put some TLC into it to make sure I don't break anything.

 I imagine the 0.7 backport will be a piece of cake, but famous last words, perhaps.

**Comment by Cadair** [comment]
:+1: on the code. No clue on the actual equations!

**Comment by Cadair** [comment]
@dpshelio :+1:  / :-1: ??

**Comment by dpshelio** [comment]
:+1:  - but just warn the people to take these calculation with a punch of salt...

**Pull request merged**